### PR TITLE
network: make packet proxy create packetconn lazily on first write

### PR DESCRIPTION
**Pull request by fortuna**
# Lazy PacketConn Initialization

Implemented lazy initialization of `net.PacketConn` in `PacketListenerProxy` and added an important `SetReadDeadline` call. The connection is now created on the first `WriteTo` call rather than eagerly in `NewSession`.

I want to pull this into https://github.com/OutlineFoundation/outline-apps/pull/2751.

#### [packet_listener_proxy.go](file:///Users/fortuna/code/Outline/outline-sdk/network/packet_listener_proxy.go)

- **Init Under Lock**: The `PacketConn` is created lazily inside `getProxyConnLocked()`, which is called by `WriteTo` while holding the `s.mu` lock.
- **Timeout**: Uses `context.WithTimeout` with `writeIdleTimeout` for the `ListenPacket` call.
- **`SetReadDeadline`**: Refreshed on every `WriteTo` to ensure the background response-relay goroutine eventually exits if the remote side stops sending data, preventing resource leaks.
- **Data Race Fix**: Protected the `writeIdleTimer` assignment in `NewSession` with a lock and added a comment explaining that it prevents a race condition with the timer goroutine calling `Close()`.

#### [packet_listener_proxy_test.go](file:///Users/fortuna/code/Outline/outline-sdk/network/packet_listener_proxy_test.go)

- **Comprehensive Revamp**: Upgraded mocks and added robust tests for all state transitions, idle/setup timeouts, and concurrency corner cases.

### Automated Tests
- All tests pass with race detection enabled.

### CODE REVIEW

File: network/packet_listener_proxy.go [hunk 1]
```diff
@@ -124,47 +167,62 @@ func (proxy *PacketListenerProxy) NewSession(respWriter PacketResponseReceiver)
...
-			if _, err := respWriter.WriteFrom(buf[:n], srcAddr); err != nil {
+			if _, err := s.respReceiver.WriteFrom(buf[:n], srcAddr); err != nil {
 				return
 			}
 		}
 	}()
 
-	return reqSender, nil
+	return nil
 }
```
[the remark this answers is not in the document]
Comment (fortuna) [reply network/packet_listener_proxy.go:182]: This is a good point I actually considered, but running the ListenPacket outside the lock adds significant complexity (e.g. an "initializing" state) that I ended up deciding to revert

I'll also note that this is not worse that the existing code, where the initialization already blocks everything else.

File: network/packet_listener_proxy.go [hunk 2]
```diff
@@ -88,27 +113,45 @@ func WithPacketListenerWriteIdleTimeout(timeout time.Duration) func(*PacketListe
...
+		return nil
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), s.writeIdleTimeout)
+	defer cancel()
+
+	proxyConn, err := s.listener.ListenPacket(ctx)
+	if err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (fortuna) [reply network/packet_listener_proxy.go:154]: Same comment as Graptile, already addressed

File: network/packet_listener_proxy.go [hunk 3]
```diff
@@ -124,47 +167,62 @@ func (proxy *PacketListenerProxy) NewSession(respWriter PacketResponseReceiver)
...
 
 // WriteTo implements [PacketRequestSender].WriteTo function. It simply forwards the packet to the underlying
 // [net.PacketConn].WriteTo function.
 func (s *packetListenerRequestSender) WriteTo(p []byte, destination netip.AddrPort) (int, error) {
-	if err := s.resetWriteIdleTimer(); err != nil {
+	s.mu.Lock()
+	if s.closed {
+		s.mu.Unlock()
+		return 0, ErrClosed
+	}
```
[the remark this answers is not in the document]
Comment (fortuna) [reply network/packet_listener_proxy.go:191]: We should be able to send empty packets. I changed the PacketRequestSender contract.

File: network/packet_listener_proxy.go [hunk 4]
```diff
@@ -88,27 +113,45 @@ func WithPacketListenerWriteIdleTimeout(timeout time.Duration) func(*PacketListe
...
 
-	// Terminate the session after timeout with no outgoing writes (deadline is refreshed by WriteTo)
+	// Terminate the session after timeout with no outgoing writes (deadline is refreshed by WriteTo).
+	// We lock here because the timer goroutine could call Close() and access writeIdleTimer concurrently.
+	reqSender.mu.Lock()
 	reqSender.writeIdleTimer = time.AfterFunc(reqSender.writeIdleTimeout, func() {
 		reqSender.Close()
 	})
+	reqSender.mu.Unlock()
+
```
[the remark this answers is not in the document]
Comment (fortuna) [reply network/packet_listener_proxy.go:135]: This was fixed by resetting the timeout just before Listen.

File: network/packet_listener_proxy_test.go [hunk 5]
```diff
@@ -35,3 +42,422 @@ func TestWithWriteTimeoutOptionWorks(t *testing.T) {
...
+	for i := 0; i < 50; i++ {
+		wg.Add(1)
+		go func(val int) {
+			defer wg.Done()
+			_, err := sender.WriteTo([]byte("data"), netip.MustParseAddrPort("127.0.0.1:1234"))
+			require.NoError(t, err)
+		}(i)
+	}
+
+	wg.Wait()
```
[the remark this answers is not in the document]
Comment (fortuna) [reply network/packet_listener_proxy_test.go]: Fixed

File: network/packet_listener_proxy.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (fortuna) [reply network/packet_listener_proxy.go:182]: Here is an explanation of the state machine for the `packetListenerRequestSender`, covering both the **current implementation** and how the **proposed fix** evolves it.

### Current State Machine

Currently, the session operates in one of three states, governed by two variables: `closed` and `proxyConn` (protected by `s.mu`).

```mermaid
stateDiagram-v2
    [*] --> Uninitialized: NewSession()
    
    Uninitialized --> Active: First successful WriteTo\n(ListenPacket OK)
    Uninitialized --> Closed: Close() or Idle Timeout
    Uninitialized --> Uninitialized: WriteTo Fails\n(ListenPacket Error)
    
    Active --> Closed: Close() or Idle Timeout
    
    Closed --> [*]
```

#### 1. Uninitialized (`!closed`, `proxyConn == nil`)
*   **Entry:** Initial state after `NewSession()`.
*   **Status:** No OS resources (sockets, goroutines) are allocated yet. The `writeIdleTimer` is running.
*   **Transitions:**
    *   To **Active**: When the first `WriteTo` successfully creates the socket via `ListenPacket`.
    *   To **Closed**: If `Close()` is called or the idle timer expires before any data is sent.
    *   Stays **Uninitialized**: If `ListenPacket` fails (e.g., temporary network error), allowing subsequent `WriteTo` calls to retry.

#### 2. Active (`!closed`, `proxyConn != nil`)
*   **Status:** The UDP socket is open, and the background goroutine is actively relaying responses.
*   **Transitions:**
    *   To **Closed**: On explicit `Close()` or if no data is sent for the `writeIdleTimeout` duration.

#### 3. Closed (`closed == true`)
*   **Status:** Terminal state. Sockets are closed, timers stopped, and resources freed.
*   **Behavior:** All future `WriteTo` or `Close` calls immediately return `ErrClosed`.

---

### Proposed State Machine (With Non-Blocking Fix)

The proposed fix introduces an **Initializing** state to prevent holding the mutex lock during the slow `ListenPacket` call.

```mermaid
stateDiagram-v2
    [*] --> Uninitialized: NewSession()
    
    Uninitialized --> Initializing: First WriteTo (Triggers ListenPacket)
    Uninitialized --> Closed: Close() or Idle Timeout
    
    Initializing --> Active: ListenPacket Succeeds
    Initializing --> Uninitialized: ListenPacket Fails (Allows Retry)
    Initializing --> Closed: Close() / Idle Timeout (Aborts ListenPacket)
    
    Active --> Closed: Close() or Idle Timeout
    
    Closed --> [*]
```

#### The new "Initializing" State (`!closed`, `proxyConn == nil`, `initCh != nil`)
*   **Why it's needed:** Allows the thread calling `ListenPacket` to drop the mutex lock so other operations aren't blocked.
*   **Behavior:**
    *   **Concurrent Writes:** If other goroutines call `WriteTo` while in this state, they detect `initCh != nil` and block on `<-initCh` rather than locking the mutex.
    *   **Cancellation:** If `Close()` is called, it immediately acquires the mutex, sets `closed = true`, and calls `cancelInit()` to abort the in-flight network call.

This addition resolves the P2 latency concern where a slow or hanging `ListenPacket` could freeze the entire proxy session management.

[the remark this answers is not in the document]
Comment (fortuna) [reply network/packet_listener_proxy.go:182]: Notes:

- The calls to WriteTo would all still block on the ListenPacket. No gains there.
- The only advantage is if we call Close() during ListenPacket, which would allow the WriteTo to return and the receiver to be closed sooner. But ListenPacket already has a deadline, so I wouldn't worry too much about it.
- Perhaps it helps at shutdown time, when we tear down the VPN. This wouldn't block.

File: network/packet_listener_proxy.go [hunk 6]
```diff
@@ -88,27 +113,49 @@ func WithPacketListenerWriteIdleTimeout(timeout time.Duration) func(*PacketListe
...
+func (s *packetListenerRequestSender) getProxyConnLocked() error {
+	if s.proxyConn != nil {
+		return nil
+	}
+
+	// Reset the timer before the slow ListenPacket call to prevent it from firing
+	// while we are blocked here.
+	s.writeIdleTimer.Reset(s.writeIdleTimeout)
+
```
[the remark this answers is not in the document]
Comment (fortuna) [reply network/packet_listener_proxy.go:147]: That's fine. It's the natural race against the timeout clock.

File: network/packet_listener_proxy.go [hunk 7]
```diff
@@ -124,47 +171,61 @@ func (proxy *PacketListenerProxy) NewSession(respWriter PacketResponseReceiver)
...
+	if err := s.getProxyConnLocked(); err != nil {
+		s.mu.Unlock()
 		return 0, err
 	}
-	return s.proxyConn.WriteTo(p, net.UDPAddrFromAddrPort(destination))
+
+	// Refresh the idle timer and read deadline.
+	s.writeIdleTimer.Reset(s.writeIdleTimeout)
+	s.proxyConn.SetReadDeadline(time.Now().Add(s.writeIdleTimeout))
+
```
[the remark this answers is not in the document]
Comment (fortuna) [reply network/packet_listener_proxy.go:201]: If a WriteTo call arrives at the exact millisecond the timer expires, the callback could already be queued and will close the session right after the write succeeds.

However, fully guarding against this would require adding extra state tracking (such as a generation ID or a lastWriteTime timestamp) to validate the callback under the lock. Given that this only impacts packets arriving at the absolute boundary of the 30-second timeout, the practical impact is minuscule.

I've decided to accept this tradeoff in favor of keeping the session lifecycle logic simpler and avoiding the added maintenance overhead.
